lib/sync.js

```diff
@@ -0,0 +1,103 @@
+var md5 = require('create-hash/md5')
+var rmd160 = require('ripemd160')
+var sha = require('sha.js')
+
+var checkParameters = require('./precondition')
+var defaultEncoding = require('./default-encoding')
+var Buffer = require('safe-buffer').Buffer
+var ZEROS = Buffer.alloc(128)
+var sizes = {
+  md5: 16,
+  sha1: 20,
+  sha224: 28,
+  sha256: 32,
+  sha384: 48,
+  sha512: 64,
+  rmd160: 20,
+  ripemd160: 20
+}
+function computePad (alg, key) {
+  var hash = getDigest(alg)
+  var blocksize = (alg === 'sha512' || alg === 'sha384') ? 128 : 64
+  if (key.length > blocksize) {
+    key = hash(key)
+  } else if (key.length < blocksize) {
+    key = Buffer.concat([key, ZEROS], blocksize)
+  }
+  var ipad = Buffer.allocUnsafe(blocksize)
+  var opad = Buffer.alloc(blocksize + sizes[alg])
+
+  for (var i = 0; i < blocksize; i++) {
+    ipad[i] = key[i] ^ 0x36
+    opad[i] = key[i] ^ 0x5C
+  }
+  return {
+    ipad: ipad,
+    opad: opad,
+    alg: alg,
+    blocksize: blocksize,
+    hash: hash
+  }
+}
+function hmac (pad, data) {
+  var h = pad.hash(Buffer.concat([pad.ipad, data]))
+  h.copy(pad.opad, pad.blocksize)
+  return pad.hash(pad.opad)
+}
+function getDigest (alg) {
+  if (alg === 'rmd160' || alg === 'ripemd160') {
+    return rmd160
+  }
```

Comment:
actually in this case since it always does it in one go there wouldn't be much benefit, I'd ideally like to upgrade hmac and hash at the same time, and maybe rewrite the md5 stuff to also be streaming so we could use a single api